
# The first rule in the document contains a list of all the deadend outputs expected
# from the pipeline, that is, every output that isn't used in a subsequent rule.
configfile:"species-specific-configs/Base_config.yaml"

rule all:
    input:
        config["species"]+"/rRNA/rRNA_index/rRNA_bowtie.1.ebwt"


# rDNA
rule download_rRNA:
    threads:
        1
    params:
        rRNA=config["rDNA_sequence"],
        outdir=config["species"]+"/rRNA/rRNA_sequences",
        cluster= '-N 1 -c 1 --mem=16G -t 00:30:00'
    output:
        fasta=config["species"]+"/rRNA/rRNA_sequences/rRNA.fa"
    shell:
        """
        cd {params.outdir}
        datasets download gene symbol {params.rRNA}
        unzip ncbi_dataset.zip
        mv ncbi_dataset/data/rna.fna rRNA.fa
        rm -r ncbi_dataset.zip ncbi_dataset README.md
        """

rule rRNA_bowtie_index:
    input:
        config["species"]+"/rRNA/rRNA_sequences/rRNA.fa"
    threads:
        4
    params:
        outdir=config["species"]+"/rRNA/rRNA_index/rRNA_bowtie",
        cluster= '-N 1 -c 1 --mem=16G -t 00:30:00',
        log=config["species"]+"/logs"
    output:
        index=config["species"]+"/rRNA/rRNA_index/rRNA_bowtie.1.ebwt"
    shell:
        """
        bowtie-build {input} 1> {params.outdir} 2> {params.log}/rRNA_bowtie.log
        """



# snRNA
# canonical snRNAs
# mature snRNAs all
# immature snRNA (genomic flanks)

# tRNA



# chrM

# genome minus chrM

